### PR TITLE
Improvement: Skip validation checks

### PR DESCRIPTION
**Pull request by maxelliott-causalens**
## Description and Motivation

Added the `validate` flag to `cai_causal_graph.causal_graph.CausalGraph.add_edge`,
  `cai_causal_graph.causal_graph.CausalGraph.add_edges_from`,
  `cai_causal_graph.causal_graph.CausalGraph.add_edges_from_paths`,
  `cai_causal_graph.causal_graph.CausalGraph.add_edge_by_pair`,
  `cai_causal_graph.causal_graph.CausalGraph.from_dict`,
  `cai_causal_graph.time_series_causal_graph.TimeSeriesCausalGraph.add_edge`, and
  `cai_causal_graph.time_series_causal_graph.TimeSeriesCausalGraph.add_time_edge` which, if set to `False`, will disable validation checks. Currently, this disables cyclicity checks when adding edges. Default is `True`, making the change backwards compatible. There is no guarantees about the behavior of the resulting graph if this is disabled specifically to introduce cycles. This should only be used to speed up this method in situations where it is known the resulting graph is still valid, for example when copying a graph.

## Additional Context

For more efficient graph extensions and copying. See the [benchmarking results here](https://docs.google.com/document/d/1yfhlruaaPf043ZRNo-7ix4cHk013rPhDvgGQMt0iWhs/edit) for benefits of skipping these checks.
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Improvement (non-breaking change which improves functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactor (moving code around, general refactoring improvements)
- [ ] Documentation (documentation)

## PR Checklist:

- [x] I have reviewed my own PR? (review the PR as you are reviewing someone else's PR)
- [x] I have implemented all requirements? (see JIRA, project documentation)
- [ ] I am affecting someone else's work? If so: I have added those people as reviewers?
- [ ] I have added comments to all the bits that are hard to follow
- [x] At a bare minimum, I tested this manually
- [x] I have added unit test(s)
- [ ] Is this a bugfix? If so have I added a regression test?
- [x] Does this PR satisfy the [one PR, one concern](https://medium.com/@fagnerbrack/one-pull-request-one-concern-e84a27dfe9f1) rule?
- [ ] If needed, documentation has been added/updated?
- [x] I have updated the appropriate changelog with a line for my changes
- [ ] If this PR breaks reproducibility, have I added a note in the changelog explaining the break in reproducibility
      and its extent?

### CODE REVIEW

File: cai_causal_graph/causal_graph.py [hunk 1]
```diff
@@ -1947,7 +1973,7 @@ def copy(self, include_meta: bool = True) -> CausalGraph:
         :return: A copy of the `cai_causal_graph.causal_graph.CausalGraph` instance.
         """
         graph_dict = self.to_dict(include_meta=include_meta)
-        new_graph = self.__class__.from_dict(graph_dict)
+        new_graph = self.__class__.from_dict(graph_dict, validate=False)
```
Comment (andrew-causalens) [inline cai_causal_graph/causal_graph.py:1976]: Why `False`? One could have a graph with cycles but I guess copy is fine with that.

File: cai_causal_graph/time_series_causal_graph.py [hunk 2]
```diff
@@ -245,7 +245,8 @@ def get_minimal_graph(self) -> TimeSeriesCausalGraph:
 
                 edge = self._EdgeCls(source=source, destination=destination, edge_type=edge.edge_type, meta=edge.meta)
 
-                minimal_cg.add_edge(edge=edge)
+                # Do not check for acyclicity as we know the original graph was a DAG, so this will be a DAG
```
Comment (andrew-causalens) [inline cai_causal_graph/time_series_causal_graph.py:248]: Do we definitely know this? Should we check it is a DAG? One can make a TS graph that is not a DAG.

File: cai_causal_graph/time_series_causal_graph.py [hunk 3]
```diff
@@ -275,8 +276,8 @@ def get_minimal_graph(self) -> TimeSeriesCausalGraph:
                     edge = self._EdgeCls(
                         source=source, destination=destination, edge_type=edge.edge_type, meta=edge.meta
                     )
-
-                    minimal_cg.add_edge(edge=edge)
+                    # Do not check for acyclicity as we know the original graph was a DAG, so this will be a DAG
```
Comment (andrew-causalens) [inline cai_causal_graph/time_series_causal_graph.py:279]: Do we definitely know this? Should we check it is a DAG? One can make a TS graph that is not a DAG.

File: changelog.md [hunk 4]
```diff
@@ -1,5 +1,14 @@
 # Changelog
 
+## 0.4.1
+
+- Added the `validate` flag to `cai_causal_graph.causal_graph.CausalGraph.add_edge`,
+  `cai_causal_graph.causal_graph.CausalGraph.add_edges_from`,
+  `cai_causal_graph.causal_graph.CausalGraph.add_edges_from_paths`,
+  `cai_causal_graph.causal_graph.CausalGraph.add_edge_by_pair`, and
+  `cai_causal_graph.causal_graph.CausalGraph.from_dict`, which if set to `False` will disable validation checks.
+  Currently, this disables cyclicity checks when adding edges.
```
Comment (andrew-causalens) [inline changelog.md:10]: Would mention default is `True` to maintain backwards compatability.

File: cai_causal_graph/time_series_causal_graph.py [hunk 5]
```diff
@@ -504,6 +505,7 @@ def extend_graph(
                             destination=lagged_destination_node,
                             edge_type=edge.get_edge_type(),
                             meta=edge.meta,
+                            validate=False,
```
Comment (andrew-causalens) [inline cai_causal_graph/time_series_causal_graph.py:508]: Are we safe doing this? I guess if there is a cycle that is OK as we want it in the extended graph. For example a valid minimal graph is X[t-1] -> X[t] -> Y[t] -> Z[t] -> X[t] if we allow cycles, which one can make by forcing that `validate=False` when adding Z[t] -> X[t] edge. 

We would at least need some unit tests for this now.  For CG and TSCG for what happens once someone has a graph in a state where they ignore validation. Not sure we are confident with how it will all work.

File: cai_causal_graph/time_series_causal_graph.py [hunk 6]
```diff
@@ -554,6 +556,7 @@ def extend_graph(
                         destination=extended_graph.get_node(lagged_dest.identifier),
                         edge_type=edge.get_edge_type(),
                         meta=edge.meta,
+                        validate=False,
```
Comment (andrew-causalens) [inline cai_causal_graph/time_series_causal_graph.py:559]: Same comment. Are we safe to do this?

**Comment by andrew-causalens** [comment]
This tag should have been a dev one and not a stable. I need to revoke `0.4.1` from `pypi` now as I am not confident this will work well if someone builds a graph purposely adding edges to create cycles by setting `validate=False`. What happens as this graph now when converting to other types and getting minimal and extended graph when it is TSCG.

**Comment by andrew-causalens** [comment]
> This tag should have been a dev one and not a stable. I need to revoke `0.4.1` from `pypi` now as I am not confident this will work well if someone builds a graph purposely adding edges to create cycles by setting `validate=False`. What happens as this graph now when converting to other types and getting minimal and extended graph when it is TSCG.

Yanked it: https://pypi.org/project/cai-causal-graph/#history

File: cai_causal_graph/time_series_causal_graph.py (hunk 2 above)
Comment (maxelliott-causalens) [reply cai_causal_graph/time_series_causal_graph.py:248]: True, I've updated the comment to be more accurate ("Do not check for acyclicity as an acyclic minimal graph will always produce an acyclic extended graph")

File: cai_causal_graph/time_series_causal_graph.py (hunk 5 above)
Comment (maxelliott-causalens) [reply cai_causal_graph/time_series_causal_graph.py:509]: Have added a warning in the docstrings that this shouldn't be done to introduce cycles, and we don't guarantee any behavior if someone does this.

File: cai_causal_graph/time_series_causal_graph.py (hunk 6 above)
Comment (maxelliott-causalens) [reply cai_causal_graph/time_series_causal_graph.py:560]: See above

File: cai_causal_graph/causal_graph.py [hunk 7]
```diff
@@ -1839,8 +1871,15 @@ def to_gml_string(self) -> str:
...
-    def from_dict(cls, d: dict) -> CausalGraph:
-        """Construct a `cai_causal_graph.causal_graph.CausalGraph` instance from a Python dictionary."""
+    def from_dict(cls, d: dict, validate: bool = True) -> CausalGraph:
+        """
+        Construct a `cai_causal_graph.causal_graph.CausalGraph` instance from a Python dictionary.
+
+        :param d: Dictionary to build a graph from.
+        :param validate: Whether to perform validation checks. The validation checks will raise if
+            any cycles are introduced to the graph by adding the edge. This should only be disabled to speed up
+            this method in situations where it is known that the serialized graph is acyclic. Default is `True`.
```
Comment (andrew-causalens) [inline cai_causal_graph/causal_graph.py:1881]: A <-> B -> C is not really acyclic. Would say "... that the serialized graph is valid." instead

File: cai_causal_graph/time_series_causal_graph.py (hunk 5 above)
Comment (andrew-causalens) [reply cai_causal_graph/time_series_causal_graph.py:509]: Would just add a comment here so we remember why we are setting it to `False`.

File: cai_causal_graph/time_series_causal_graph.py (hunk 6 above)
Comment (andrew-causalens) [reply cai_causal_graph/time_series_causal_graph.py:560]: Would just add a comment here so we remember why we are setting it to `False`.

File: cai_causal_graph/time_series_causal_graph.py [hunk 8]
```diff
@@ -964,7 +987,7 @@ def from_causal_graph(cls, causal_graph: CausalGraph) -> TimeSeriesCausalGraph:
             assert isinstance(source, TimeSeriesNode)  # for linting
             assert isinstance(destination, TimeSeriesNode)  # for linting
 
-            ts_cg.add_edge(source, destination, meta=edge.meta, edge_type=edge.get_edge_type())
+            ts_cg.add_edge(source, destination, meta=edge.meta, edge_type=edge.get_edge_type(), validate=False)
```
Comment (andrew-causalens) [inline cai_causal_graph/time_series_causal_graph.py:990]: Would just add a comment here so we remember why we are setting it to `False`.

File: changelog.md [hunk 9]
```diff
@@ -1,5 +1,17 @@
 # Changelog
 
+## 0.4.1
+
+- Added the `validate` flag to `cai_causal_graph.causal_graph.CausalGraph.add_edge`,
```
Comment (andrew-causalens) [inline changelog.md:5]: Missing the TS graph public methods that this was added to.

File: changelog.md [hunk 10]
```diff
@@ -1,5 +1,19 @@
 # Changelog
 
+## 0.4.1
```
Comment (andrew-causalens) [inline changelog.md:3]: Should be NEXT.

**Review by andrew-causalens — APPROVED** [review]
